### ISSUE TITLE
Maven proxy

### ISSUE THREAD

**Issue by g4s8** [artipie/maven-adapter#92]
Implement Maven proxy repository (or Maven mirror) with authentication and cache support. Maven mirrors are described here: https://maven.apache.org/guides/mini/guide-mirror-settings.html

## Multi-remote

Maven proxy can be used to merge multiple remote repositories into single one. In this case, proxy repository is trying to access remotes one by one in specified order until find applicable artifact.

*Example: user uses Artipie Maven mirror to get artifacts from custom private repository and from Maven central, user configures Artipie Maven mirror to first search artifacts in private repository, and then from central, if artifact was not found in private repo.*

## Authentication

Proxy should be able to use authentication to access remote repository. The use case is when the user is giving access to different repositories through proxy or hiding actual credentials on proxy side.

## Caching

Proxy should be able to cache artifacts and metadata, the user can configure the TTL and size of the cache. Proxy needs to check HTTP cache control directives when pulling remote artifacts.
Artipie Maven proxy should use cache directives as well to avoid redundant downloads.

### Configuration

The suggested configuration for proxy repository for Artipie is:
```yaml
repository:
  type: maven-proxy
  remotes:
    - url: https://private.repo.com/maven # required
      username: admin # optional
      password: qwerty # optional
      cache: # optional
        ttl: 5d # optional
        maxSize: 10G # optional
    - url: https://repo.maven.apache.org/maven2
```

This job may be discussed and changed, also it can be boosted if needed.

**Comment by 0crat** [comment]
The job [#92](https://github.com/artipie/maven-adapter/issues/92) assigned to @victornoel[/z](https://www.0crat.com/u/victornoel), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/9ad0b00e-966f-498d-9621-bb6ec3c90172); the budget is 30 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§8](http://www.zerocracy.com/policy.html#8) and [§9](http://www.zerocracy.com/policy.html#9); if the task is not clear, read [this](http://www.yegor256.com/2015/02/16/it-is-not-a-school.html) and [this](http://www.yegor256.com/2015/01/15/how-to-cut-corners.html); there will be a monetary reward for this job

**Comment by victornoel** [comment]
@g4s8 So I was thinking about the following design
- one `MavenSlice`
  - protected as it currently is
- multiple `MavenRepository` client implementation
  - `LocalMavenRepository` would be the current behaviour looking for the files locally
  - `RemoteMavenRepository` would be getting artefacts from a remote repository
    - it will also manage auth and cache
  - `MavenRepositories` that would handle choosing from one of the repositories it knows
    - First implementation would be to simply to have an ordered list of repository and look at them in order as you propose in the description
    - Later we can imagine other policies for choosing the remote repository to use
    - We should be careful about being deterministic with respect to serving ALL the files (metadata, jar, pom, etc) of a given artifact from one and only one remote repository, I'm not sure how solve that though yet.

Other notes:
- Is the cache related to the front-facing proxy (i.e., `MavenRepositories`) or to each `RemoteRepository`?
- Should the cache be implemented simply by delegating to a `LocalMavenRepository`? Just an idea...

WDYT?

**Comment by victornoel** [comment]
@0crat wait for design to be finished

**Comment by g4s8** [comment]
@victornoel thanks, sounds good. Just a few questions:
 - What is the `*Repository` in your example (local or remote), is it a subtype of `Slice` or some new abstraction?
 - "> We should be careful about being deterministic with respect to serving ALL the files" - as I know, standard `mvn` client uses this policy: it may have multiple repository configurations in `settings.xml` and it's looking for dependencies in each repository until it finds any
 - "> Is the cache related to the front-facing proxy? > Should the cache be implemented simply by delegating to a LocalMavenRepository?" - I think caching is a bit more complex: remote proxy cache should obey disk size and TTL (time to live) configuration rules and remote cache directives (etags and max-age), but local repository doesn't do that - it doesn't have such configurations and cache directives, it's just keeping all artifacts forever or until deleted.
 - What do you think about metadata files for proxy repository? I see two possible ways: download remote metadata files and push to cache without touching it (just handle as other artifact binaries), or update it on proxy level? It may be not easy question, since proxy may contain metadata files with all list of artifact versions, e.g. proxy serves metadata with `1.0`, `2.0`, `3.0` versions, but keeps only `1.0` artifact version. And we need to synchronize metadata files between proxy and remote repository, maybe just update it on GET request to proxy.

**Comment by victornoel** [comment]
@g4s8 

> What is the `*Repository` in your example (local or remote), is it a subtype of `Slice` or some new abstraction?

I was thinking a new abstraction, but made to have the `MavenSlice` as its only user, so it is providing behaviour for retrieving files and `MavenSlice` will delegate to it.

>> We should be careful about being deterministic with respect to serving ALL the files

> as I know, standard mvn client uses this policy: it may have multiple repository configurations in settings.xml and it's looking for dependencies in each repository until it finds any

That's not what I am talking about, I'm talking about preventing a situation where the jar is retrieved from one of the remote repository and the pom is retrieved from another one.

> What do you think about metadata files for proxy repository? I see two possible ways: download remote metadata files and push to cache without touching it (just handle as other artifact binaries), or update it on proxy level? It may be not easy question, since proxy may contain metadata files with all list of artifact versions, e.g. proxy serves metadata with 1.0, 2.0, 3.0 versions, but keeps only 1.0 artifact version. And we need to synchronize metadata files between proxy and remote repository, maybe just update it on GET request to proxy.

I'm not sure, it's complex. I suppose we could look at how other proxy do, such as Nexus. I think they maintain some kind of metadata for the proxy that is the aggregation of the all the metadata of all the remotes. As for the way they practically do it, maybe they do it on the fly?

What is clear is that we should first decide which behaviour we want, and then later we can see how to implement it.

**Comment by g4s8** [comment]
@victornoel 
> That's not what I am talking about, I'm talking about preventing a situation where the jar is retrieved from one of the remote repository and the pom is retrieved from another one.

Got it, I think we can try to load the whole artifact content only from one remote source.

> As for the way they practically do it, maybe they do it on the fly?

Yes we can start with the simplest possible way.

> What is clear is that we should first decide which behaviour we want, and then later we can see how to implement it.

Agree, what part is not clear now? (I mean in global feature scope)

**Comment by victornoel** [comment]
@g4s8 
> Agree, what part is not clear now? (I mean in global feature scope)

now it's clear, I was confused I think :) I will make a PR with the new interfaces/classes and todos. Can we boost the job maybe?

**Comment by g4s8** [comment]
@victornoel yes, sure, thanks.

**Comment by g4s8** [comment]
@0crat boost 4x

**Comment by 0crat** [comment]
> @0crat boost 4x ([here](https://github.com/artipie/maven-adapter/issues/92#issuecomment-637347392))

@g4s8 Boost 4x was set for [#92](https://github.com/artipie/maven-adapter/issues/92)

**Comment by g4s8** [comment]
@rultor release, tag=`0.4`

**Comment by victornoel** [comment]
@g4s8 see #94 :)

**Comment by g4s8** [comment]
@0crat status

**Comment by 0pdd** [comment]
@g4s8 4 puzzles [#95](https://github.com/artipie/maven-adapter/issues/95), [#96](https://github.com/artipie/maven-adapter/issues/96), [#97](https://github.com/artipie/maven-adapter/issues/97), [#98](https://github.com/artipie/maven-adapter/issues/98) are still not solved.

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/artipie/maven-adapter/issues/92#issuecomment-640748307))

@g4s8 This is what I know about this job in CT2E6TK9B, as in [§32](http://www.zerocracy.com/policy.html#32):

  * The job [#92](https://github.com/artipie/maven-adapter/issues/92) is [in scope](http://www.zerocracy.com/policy.html#14) for 10days
  * The role is `DEV`
  * The job is assigned to @victornoel[/z](https://www.0crat.com/u/victornoel) for [10days](http://www.zerocracy.com/policy.html#8)
  * There is a monetary reward attached
  * @victornoel[/z](https://www.0crat.com/u/victornoel) will get 100% of their hourly rate for this job
  * The job has an [impediment](http://www.zerocracy.com/policy.html#9)
  * The [budget](http://www.zerocracy.com/policy.html#4) is 60 minutes/points
  * These users are banned and won't be assigned:
    * @g4s8[/z](https://www.0crat.com/u/g4s8): This user reported the ticket
  * Job [footprint](https://www.0crat.com/footprint/CT2E6TK9B?q=%7Bjob%3A%27gh%3Aartipie%2Fmaven-adapter%2392%27%2Ctype%3A%7B%24not%3A%2F%5ENotify%2F%7D%7D) (restricted area)

**Comment by g4s8** [comment]
@victornoel thanks, closing.

**Comment by 0crat** [comment]
@sereshqua[/z](https://www.0crat.com/u/sereshqua) please review this job completed by @victornoel[/z](https://www.0crat.com/u/victornoel), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by sereshqua** [comment]
@0crat quality good

**Comment by 0pdd** [comment]
@g4s8 3 puzzles [#95](https://github.com/artipie/maven-adapter/issues/95), [#97](https://github.com/artipie/maven-adapter/issues/97), [#98](https://github.com/artipie/maven-adapter/issues/98) are still not solved; solved: [#96](https://github.com/artipie/maven-adapter/issues/96).

**Comment by 0pdd** [comment]
@g4s8 4 puzzles [#104](https://github.com/artipie/maven-adapter/issues/104), [#95](https://github.com/artipie/maven-adapter/issues/95), [#97](https://github.com/artipie/maven-adapter/issues/97), [#98](https://github.com/artipie/maven-adapter/issues/98) are still not solved; solved: [#96](https://github.com/artipie/maven-adapter/issues/96).

**Comment by 0pdd** [comment]
@g4s8 3 puzzles [#104](https://github.com/artipie/maven-adapter/issues/104), [#95](https://github.com/artipie/maven-adapter/issues/95), [#97](https://github.com/artipie/maven-adapter/issues/97) are still not solved; solved: [#96](https://github.com/artipie/maven-adapter/issues/96), [#98](https://github.com/artipie/maven-adapter/issues/98).

**Comment by 0pdd** [comment]
@g4s8 all 5 puzzles are solved here: [#104](https://github.com/artipie/maven-adapter/issues/104), [#95](https://github.com/artipie/maven-adapter/issues/95), [#96](https://github.com/artipie/maven-adapter/issues/96), [#97](https://github.com/artipie/maven-adapter/issues/97), [#98](https://github.com/artipie/maven-adapter/issues/98).
